### DISCUSSION TITLE
INWX DNS entry don't get updated

### DISCUSSION THREAD

**Discussion by Peronia** [qdm12/ddns-updater#650] [Q&A]
Hi,

I use 2 DNS providers, Netcup und INWX. Netcup works fine, no issues.
But with 2 different INWX accounts I have the problem, that ddns updater says the entry is successfull updated but this is not true. And after 10 mins it try it once more and so on.
Here are my config for one of these accounts:

```
{
            "provider": "inwx",
            "domain": "example.com",
            "host": "@",
            "username": "MyUsername",
            "password": "MyPassword",
            "ip_version": "ipv4"
}
```
I am using the v. 2.6 standalone. I used the same username and password as for the login via web.

Recent logs:
```
Feb 14 17:00:48 tethys ddns-updater[293034]: 2024-02-14T17:00:48+01:00 INFO Updating record [domain: example.com | host: @ | provider: inwx | ip: ipv4] to use XX.XX.XX.XX
Feb 14 17:10:48 tethys ddns-updater[293034]: 2024-02-14T17:10:48+01:00 INFO ipv4 address of example.com is OLD.XXX.OLD.XXX and your ipv4 address  is XX.XX.XX.XX
Feb 14 17:10:48 tethys ddns-updater[293034]: 2024-02-14T17:10:48+01:00 INFO Updating record [domain: example.com | host: @ | provider: inwx | ip: ipv4] to use XX.XX.XX.XX
Feb 14 17:20:48 tethys ddns-updater[293034]: 2024-02-14T17:20:48+01:00 INFO obtaining ipv4 address succeeded after 1 failed try
Feb 14 17:20:48 tethys ddns-updater[293034]: 2024-02-14T17:20:48+01:00 INFO ipv4 address of example.com is OLD.XXX.OLD.XXX and your ipv4 address  is XX.XX.XX.XX
Feb 14 17:20:48 tethys ddns-updater[293034]: 2024-02-14T17:20:48+01:00 INFO Updating record [domain: example.com | host: @ | provider: inwx | ip: ipv4] to useXX.XX.XX.XX
Feb 14 17:30:48 tethys ddns-updater[293034]: 2024-02-14T17:30:48+01:00 INFO ipv4 address of example.com is OLD.XXX.OLD.XXX and your ipv4 address  is XX.XX.XX.XX
Feb 14 17:30:49 tethys ddns-updater[293034]: 2024-02-14T17:30:49+01:00 INFO Updating record [domain: example.com | host: @ | provider: inwx | ip: ipv4] to use XX.XX.XX.XX
```

This is all from the same account where a wrong IP is set. Also in the backoffice, I can see there is no change.
What is wrong here?

**Comment by Peronia** [comment]
Okay, I added log level debug and it shows me the called url for update. It seems that the password is not that one from the config. It have a wrong encoding for special chars like ! is %21. I changed it to a password without special chars and this doesn't work too. In the answer body i found that: `body: nochg` so thats correct, no change, but I need a change.

**Comment by qdm12** [comment]
Hi there, sorry for the delay answering.

> But with 2 different INWX accounts I have the problem, that ddns updater says the entry is successfull updated but this is not true.

That's odd, the program only succeeds if it gets a response with prefix `good` or `nochg`, kind of weird inwx doesn't update/badly updates. Might be worth reaching out to them.

> It have a wrong encoding for special chars like ! is %21

The username password are prefixed to the url `https://user:password@dyndns.inwx.com/nic/update` and therefore the password is url-escaped. See https://datatracker.ietf.org/doc/html/rfc3986 `!` must be escaped, and if Inwx doesn't handle re-converting back from escaped characters to unescaped, it might be worth shooting them an email 😉 

> so thats correct, no change, but I need a change.

This is confusing, do you mean it works as expected or not??

**Comment by Peronia** [comment]
Okay, lets get this sorted.
> That's odd, the program only succeeds if it gets a response with prefix good or nochg, kind of weird inwx doesn't update/badly updates. Might be worth reaching out to them.

Yeah, as I wrote, I got `nochg` as answer, so it's ok that it changed to success BUT the IP wasn't changed, so after 5 mins the updater will try it again.
> This is confusing, do you mean it works as expected or not??

I foung the `nochg` in the body, thats wrong. The IP wasn't changed at INWX. I don't know why I got this. Before this updater I used [this](https://github.com/Peronia/Bash-INWX-DynDNS) script. Just tested it -> this updater fails and the "old" script works fine, both using the same account and the same domain (a root domain, no sub domain).

**Comment by Peronia** [comment]
Nobody can help here?

**Comment by qdm12** [comment]
I'll have a look at the script soon (Wednesday the earliest) to compare with the current code, maybe it's a different api endpoint or something like that now 🤔 This is one of my browser tabs (15 total 😄) so I shouldn't forget! (feel free to spam remind me though)

**Comment by qdm12** [comment]
In the page https://www.inwx.com/en/offer/dyndns they mention an update url, can you try finding what's the one for you? That's what ddns-updater uses. Please try in your browser `https://user:password@dyndns.inwx.com/nic/update?hostname=domain.com&myip=127.0.0.1` to check if this works as intended, and, if not, please reach out to inwx's support since it might simply be broken on their side.

The script you mentioned uses the API (https://www.inwx.com/en/help/apidoc) to update the record, I can do that too if the update url method still doesn't work, but this would take me a few hours and I'm kind of short on time right now.

**Comment by Peronia** [comment]
Well, I don't think this work how you think it should. We have 2 issues:

1. When I use your way, the documentation from [INWX ](https://github.com/qdm12/ddns-updater/blob/master/docs/inwx.md)is not complete. The password and username differ from the account ones.
2. When I use the correct password and username, it works. But the IP change only get to the subdomain you have configured at INWX for that. Here is the config screen:
![grafik](https://github.com/qdm12/ddns-updater/assets/25049991/e672cccd-399a-435c-ba5f-24b6f151be60)
So ddns.example.com get the new IP, but not example.com. And here we have the problem. DDNS updater checks example.com and trigger the update once more...

**Reply by qdm12** [reply]
Ah ok this makes sense. Do you think it's better to:

1. fix the documentation and have credentials per full hostname (same code as now)
2. change the code to use their API and use the account credentials to update (and create if it doesn't exist, see #129). This one may break compatibility if someone figured out how to use the existing code correctly though.

**Reply by Peronia** [reply]
I don't think that 1. is the solution. It is only a solution when:

- You use only a subdomain for DDNS, not an entry like @
- You use a subdomain for DDNS and your @ entry is a cname to that subdomain

The first one is possible but the second one where really uncommon, also INWX inform you about that when you change the @ record to cname.
So I think we should really use the API, maybe you can have a look at my script mentioned before to waste not to much time. When you configure a DDNS account at INWX, you MUST use a subdomain, entries like @ doesn't work (I don't tested it, but at the picture you see subdomain, so I would assume it).
Now about 2.:
Main people are using @ for the IP to update, why we need such a feature? I'm confused, it's a feature but I don't know how important that is. And lastly I don't understand your complain about compatibility.

**Reply by Peronia** [reply]
Ah, I've now realized what you meant by compatibility. How about a new DNS provider, like INWX_API?

**Reply by Peronia** [reply]
@qdm12 (:
